src/Kestrel.Core/Internal/Http2/Http2FrameWriter.cs

```diff
@@ -109,51 +117,132 @@ public void WriteResponseHeaders(int streamId, int statusCode, IHeaderDictionary
+                // Zero-length data frames are allowed to be sent immediately even if there is no space available in the flow control window.
+                // https://httpwg.org/specs/rfc7540.html#rfc.section.6.9.1
+                if (dataLength != 0 && dataLength > flowControl.Available)
+                {
+                    return WriteDataAsyncAwaited(streamId, flowControl, data, dataLength, endStream);
                 }
 
-                _outgoingFrame.Length = data.Length;
+                flowControl.Advance(dataLength, out _);
+                return WriteDataUnsynchronizedAsync(streamId, data, endStream);
+            }
+        }
+
+        private Task WriteDataUnsynchronizedAsync(int streamId, ReadOnlySequence<byte> data, bool endStream)
+        {
+            if (_completed)
+            {
+                return Task.CompletedTask;
+            }
+
+            _outgoingFrame.PrepareData(streamId);
 
-                if (endStream)
+            var payload = _outgoingFrame.Payload;
+            var unwrittenPayloadLength = 0;
+
+            foreach (var buffer in data)
+            {
+                var current = buffer;
+
+                while (current.Length > payload.Length)
+                {
+                    current.Span.Slice(0, payload.Length).CopyTo(payload);
+                    current = current.Slice(payload.Length);
+
+                    _outputWriter.Write(_outgoingFrame.Raw);
+                    payload = _outgoingFrame.Payload;
+                    unwrittenPayloadLength = 0;
+                }
+
+                if (current.Length > 0)
                 {
-                    _outgoingFrame.DataFlags = Http2DataFrameFlags.END_STREAM;
+                    current.Span.CopyTo(payload);
+                    payload = payload.Slice(current.Length);
+                    unwrittenPayloadLength += current.Length;
                 }
+            }
+
+            if (endStream)
+            {
+                _outgoingFrame.DataFlags = Http2DataFrameFlags.END_STREAM;
+            }
+
+            _outgoingFrame.Length = unwrittenPayloadLength;
+            _outputWriter.Write(_outgoingFrame.Raw);
+
+            return FlushUnsynchronizedAsync();
+        }
+
+        private async Task WriteDataAsyncAwaited(int streamId, Http2StreamFlowControl flowControl, ReadOnlySequence<byte> data, long dataLength, bool endStream)
+        {
+            while (dataLength > 0)
+            {
+                Http2FlowControlAwaitable availabilityAwaitable;
+                var writeTask = Task.CompletedTask;
+
+                lock (_writeLock)
+                {
+                    if (flowControl.IsAborted)
+                    {
+                        break;
+                    }
+
+                    var actual = flowControl.Advance(dataLength, out availabilityAwaitable);
 
-                data.CopyTo(_outgoingFrame.Payload);
+                    if (actual > 0)
```

Comment:
I expect we'll want some minimum here greater than 0 to keep from writing one byte at a time.

Reply:
E.g. ResumeWriterThreshold